File: leetcode/1402.scala
object Solution {
  def maxSatisfaction(satisfaction: Array[Int]): Int = {
    val sortedSatisfaction = satisfaction.sorted
    
    sortedSatisfaction.indices
      .foldLeft (0) {
        (max, i) => Math.max(max, sumSatisfaction(sortedSatisfaction drop i))
      }
  }

  def sumSatisfaction(satisfaction: Array[Int]): Int =
    satisfaction.zipWithIndex
      .foldLeft (0) {
        (sum, pair) => sum + (pair._2 + 1) * pair._1
      }
}

File: leetcode/1402.js
/**
 * @param {number[]} satisfaction
 * @return {number}
 */
const maxSatisfaction = (satisfaction) => {
  const increasing = (a, b) => a - b;
  const sortedSatisfaction = satisfaction.sort(increasing);
  return sortedSatisfaction.reduce(
    (max, _, i, arr) => Math.max(max, sumSatisfaction(arr.slice(i))),
    0,
  );
};

const sumSatisfaction = (satisfaction) =>
  satisfaction.reduce((sum, a, i) => sum + (i + 1) * a, 0);
